### PR TITLE
Add release notes for 1.7.0

### PR DESCRIPTION
**Pull request by straight-shoota**

### CODE REVIEW

File: _releases/2023-01-09-1.7.0-released.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+```crystal
+BigDecimal.new(Float32::MAX) # => 3.4028235e+38
+```
+
+Further additions:
+
+* Automatically cast Int to Float for `{JSON,YAML}::Any#as_f` ([#12835](https://github.com/crystal-lang/crystal/pull/12835)).
+* Several missing `#index!`/`#rindex!` methods on `String` and `Indexable`([#12730](https://github.com/crystal-lang/crystal/pull/12730), [#12759](https://github.com/crystal-lang/crystal/pull/12759)).
+* Methods to manipulate `SemanticVersion` ([#12834](https://github.com/crystal-lang/crystal/pull/12834).
+* `HTTP::Server::Response#redirect` ([#12526](https://github.com/crystal-lang/crystal/pull/12526).
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:88]:
```suggestion
* `HTTP::Server::Response#redirect` ([#12526](https://github.com/crystal-lang/crystal/pull/12526)).
```

File: _releases/2023-01-09-1.7.0-released.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+```crystal
+BigDecimal.new(Float32::MAX) # => 3.4028235e+38
+```
+
+Further additions:
+
+* Automatically cast Int to Float for `{JSON,YAML}::Any#as_f` ([#12835](https://github.com/crystal-lang/crystal/pull/12835)).
+* Several missing `#index!`/`#rindex!` methods on `String` and `Indexable`([#12730](https://github.com/crystal-lang/crystal/pull/12730), [#12759](https://github.com/crystal-lang/crystal/pull/12759)).
+* Methods to manipulate `SemanticVersion` ([#12834](https://github.com/crystal-lang/crystal/pull/12834).
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:87]:
```suggestion
* Methods to manipulate `SemanticVersion` ([#12834](https://github.com/crystal-lang/crystal/pull/12834), see also the [docs](https://crystal-lang.org/api/1.7.0/SemanticVersion.html)).
```

File: _releases/2023-01-09-1.7.0-released.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+* `HTTP::Server::Response#redirect` ([#12526](https://github.com/crystal-lang/crystal/pull/12526).
+
+# Code quality
+
+There were a couple of changes to improve code quality in the entire repository.
+Most of them are based on suggestions by the static code analyzer tool [ameba](https://github.com/crystal-ameba/ameba).
+Nothing is automated yet, so we can't ensure this for code additions. But that would one of the next steps.
+
+A nice feature supported by many code highters is context-specific heredoc deliminators. We implemented that accross the code base in [#12816](https://github.com/crystal-lang/crystal/pull/12816).
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:96]: Not sure how relevant this section is, as it doesn't affect people using the compiler (the main audience). But I don't oppose either.

```suggestion
A nice feature supported by many code highlighters is context-specific heredoc deliminators. We implemented that accross the code base in [#12816](https://github.com/crystal-lang/crystal/pull/12816).
```

File: _releases/2023-01-09-1.7.0-released.md (hunk 2 above)
Comment (straight-shoota) [reply _releases/2023-01-09-1.7.0-released.md:87]: Yeah, I haven't added all the docs links yet. I was planning to link the class name directly. Same for all the other features mentioned in the post.

File: _releases/2023-01-09-1.7.0-released.md (hunk 3 above)
Comment (straight-shoota) [reply _releases/2023-01-09-1.7.0-released.md:96]: I'm thinking of this as a lateral "did you know?" information that can certainly be helpful to many developers.

File: _releases/2023-01-09-1.7.0-released.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+# Interpreter REPL
+
+The REPL interface of the interpreter has received a major usability improvement.
+The command prompt now supports multiline input, command history, and several key bindings.
+
+[![asciicast](/assets/blog/2023-01-09-interpreter-repl-asciicast.gif)](https://asciinema.org/a/oG55o2Jm3XuNj8DJKBb9u0XuW)
+
+This great feature set is provided by the shard [REPLy](https://github.com/I3oris/reply) ([#12738](https://github.com/crystal-lang/crystal/pull/12738)).
+
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:28]:
```suggestion

For technical reasons, the interpreter currently only ships by default in `brew`. For other platforms you need to compile it from the [sources](https://crystal-lang.org/install/from_sources/) with `interpreter=1` in the `make` options.

```

File: _releases/2023-01-09-1.7.0-released.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+## Compiler
+
+Clear error message when declaring a constant within another constant declaration ([#12566](https://github.com/crystal-lang/crystal/pull/12566)).
+
+```cr
+A = B = 1 # Syntax Error: dynamic constant assignment
+```
+
+Fixed wrong/missing locations of various AST nodes in the parser ([#11798](https://github.com/crystal-lang/crystal/pull/11798)).
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:64]: Are these important to stand out here?

```suggestion
```

File: _releases/2023-01-09-1.7.0-released.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+## stdlib
+
+⚠️ [`File.real_path`](https://crystal-lang.org/api/1.7.0/File.html#real_path(path:Path|String):String-class-method) was renamed to [`File.realpath`](https://crystal-lang.org/api/1.7.0/File.html#realpath%28path%3APath%7CString%29%3AString-class-method). This spelling just makes more sense as it's used everywhere else. The old version is deprecated, but keeps working until the next major release ([#12552](https://github.com/crystal-lang/crystal/pull/12552)).
+
+⚠️ Error handling for `libXML2` does no longer leak errors outside the current context. [`XML.errors`](https://crystal-lang.org/api/1.7.0/XML/Error.html#errors:Array(XML::Error)|Nil-class-method) has been deprecated, but should continue to work. We highly recommend to access errors directly in the respective context via [`XML::Reader#errors`](https://crystal-lang.org/api/1.7.0/XML/Reader.html#errors:Array(XML::Error)-instance-method) and [`XML::Node#errors`](https://crystal-lang.org/api/1.7.0/XML/Node.html#errors:Array(XML::Error)|Nil-instance-method) ([#12663](https://github.com/crystal-lang/crystal/pull/12663), [#12795](https://github.com/crystal-lang/crystal/pull/12795)).
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:72]: Escaping
```suggestion
⚠️ [`File.real_path`](https://crystal-lang.org/api/1.7.0/File.html#real_path%28path%3APath%7CString%29%3AString-class-method) was renamed to [`File.realpath`](https://crystal-lang.org/api/1.7.0/File.html#realpath%28path%3APath%7CString%29%3AString-class-method). This spelling just makes more sense as it's used everywhere else. The old version is deprecated, but keeps working until the next major release ([#12552](https://github.com/crystal-lang/crystal/pull/12552)).

⚠️ Error handling for `libXML2` does no longer leak errors outside the current context. [`XML.errors`](https://crystal-lang.org/api/1.7.0/XML/Error.html#errors%3AArray(XML%3A%3AError)%3F-class-method) has been deprecated, but should continue to work. We highly recommend to access errors directly in the respective context via [`XML::Reader#errors`](https://crystal-lang.org/api/1.7.0/XML/Reader.html#errors%28Array%28XML%3A%3AError%29-instance-method) and [`XML::Node#errors`](https://crystal-lang.org/api/1.7.0/XML/Node.html#errors%3AArray%28XML%3A%3AError%29%7CNil-instance-method) ([#12663](https://github.com/crystal-lang/crystal/pull/12663), [#12795](https://github.com/crystal-lang/crystal/pull/12795)).
```

File: _releases/2023-01-09-1.7.0-released.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+[`BigDecimal#to_s`](https://crystal-lang.org/api/1.7.0/BigDecimal.html#to_s(io:IO):Nil-instance-method) supports scientific notation ([#10805](https://github.com/crystal-lang/crystal/pull/10805)).
+
+```crystal
+BigDecimal.new(Float32::MAX) # => 3.4028235e+38
+```
+
+Further additions:
+
+* Automatically cast Int to Float for [`JSON::Any#as_f`](https://crystal-lang.org/api/1.7.0/JSON/Any.html#as_f:Float64-instance-method) nad [`YAML::Any#as_f`](https://crystal-lang.org/api/1.7.0/YAML/Any.html#as_f:Float64-instance-method) ([#12835](https://github.com/crystal-lang/crystal/pull/12835)).
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:85]:
```suggestion
* Automatically cast from `Int` to `Float` for [`JSON::Any#as_f`](https://crystal-lang.org/api/1.7.0/JSON/Any.html#as_f:Float64-instance-method) and [`YAML::Any#as_f`](https://crystal-lang.org/api/1.7.0/YAML/Any.html#as_f:Float64-instance-method) ([#12835](https://github.com/crystal-lang/crystal/pull/12835)).
```

File: _releases/2023-01-09-1.7.0-released.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+* Several missing `#index!`/`#rindex!` methods on [`String`](https://crystal-lang.org/api/1.7.0/String.html) and [`Indexable`](https://crystal-lang.org/api/1.7.0/Indexable.html) ([#12730](https://github.com/crystal-lang/crystal/pull/12730), [#12759](https://github.com/crystal-lang/crystal/pull/12759)).
+* Methods to manipulate [`SemanticVersion`](https://crystal-lang.org/api/1.7.0/SemanticVersion.html) ([#12834](https://github.com/crystal-lang/crystal/pull/12834).
```
Comment (beta-ziliani) [inline _releases/2023-01-09-1.7.0-released.md:87]:
```suggestion
* Methods to manipulate [`SemanticVersion`](https://crystal-lang.org/api/1.7.0/SemanticVersion.html) ([#12834](https://github.com/crystal-lang/crystal/pull/12834)).
```

**Review by beta-ziliani — CHANGES_REQUESTED** [review]

**Comment by beta-ziliani** [comment]
Note: I haven't checked that my escaping is sound and complete...

File: _releases/2023-01-09-1.7.0-released.md (hunk 5 above)
Comment (straight-shoota) [reply _releases/2023-01-09-1.7.0-released.md:64]: Not really, they're on my maybe list.
Happy to drop it.

File: _releases/2023-01-09-1.7.0-released.md (hunk 4 above)
Comment (beta-ziliani) [reply _releases/2023-01-09-1.7.0-released.md:28]: What do you think of ☝️ ?

**Review by beta-ziliani — APPROVED** [review]

**Pull request merged**